zerver/actions/streams.py

```diff
@@ -428,7 +429,9 @@ def bulk_add_subscriptions(
         # Make a fresh set of all new recipient ids, and then we will
         # remove any for which our user already has a subscription
         # (and we'll re-activate any subscriptions as needed).
-        new_recipient_ids: Set[int] = {stream.recipient_id for stream in streams}
+        new_recipient_ids: Set[int] = {
+            stream.recipient_id for stream in streams if stream.recipient_id is not None
+        }
```

Comment:
This would silently drop any streams where `recipient_id` was None. That's probably not intended, right?

Reply:
Yeah, we don't want None in this set. But if we do somehow get here with a stream that has a `recipient_id` of None, I think we want to error out -- not proceed with a set from which we've silently dropped that stream.

The existing behavior is that we'll get None in the set, and then probably crash at a random spot below this line. If so, that's not as good as failing with a crisp error here -- but it likely would be better than pressing on with data that we've silently dropped some items from. That is, if we were to leave this line alone and just stick a cast or a type-ignore comment on it, that might well be better than making this change (though not as good as a version of this change that asserts out when one of these is None.)

---

At a high level, what's going on here is that the types on this Stream model (once we have such types, thanks to those spiffy new Django stubs) don't actually support the assumption this code was making, that every stream has an actual int for its `recipient_id`. That gap needs to be bridged somehow.

By just discarding the None values, this version bridges that gap basically by relaxing the assumption. But then I think that opens up a new gap -- the rest of the function was assuming that this `new_recipient_ids` contained the recipient ID of every stream in `streams`, and now it may not be. Put another way, the rest of the function was assuming that every stream has an actual recipient ID, and is still making that assumption. If we really wanted to relax that assumption, then more of this function's logic would need to be rethought with that in mind.

I think instead we want to keep that assumption. So this line should go ahead and use that assumption, either with or without a runtime check for it.

---

Reading more of the context above, it looks like you're already adding an assert that covers this, up in the `recipient_color_map` loop. So there's no need to check again, and a cast would be appropriate.

(It also seems like this code could be refactored and simplified in other ways -- e.g. this could just say `set(recipient_ids)`. But that's perhaps out of scope for this PR.)

Reply:
Thanks for the insight. `cast` might be the last resort. In this case, I think it's probably cleaner to just create the set right after the previous assertion and copy it for each `user_profile`, so we essentially leave the assumption untouched, but in a type-safe manner. I can fit this change into this PR as it will be pretty straightforward.